File: hummingbird/androidApp/app/src/main/java/net/acumenus/hummingbird/ui/flow/TransportRoutesLayer.kt
package net.acumenus.hummingbird.ui.flow

import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.clickable
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import net.acumenus.hummingbird.data.FlowFloor
import net.acumenus.hummingbird.data.FlowFloorsDocument
import net.acumenus.hummingbird.data.FlowWindowData
import net.acumenus.hummingbird.ui.components.panel
import net.acumenus.hummingbird.ui.theme.Z
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin

/** A trip endpoint resolved to the map: a floor plus an optional plan-feet anchor rect. */
data class FlowSpacePoint(val floor: Int, val rect: List<Double>?)

/**
 * One transport trip drawable on the map: a past transport_status event (solid
 * arc) or a transport_due projection still ahead of the scrub time (dashed
 * ghost). Ghost trips from derived discharges have no destination — they render
 * as departure stubs from the origin bed's unit.
 */
data class TransportTrip(
    val selection: FlowSelection,
    val label: String,
    val fromRaw: String?,
    val toRaw: String?,
    val from: FlowSpacePoint?,
    val to: FlowSpacePoint?,
    val ghost: Boolean,
    val alpha: Float,
    /** STAT/at-risk tier → warning tint; NEVER the critical color for trips. */
    val warning: Boolean,
)

/**
 * Resolves the transport from_space/to_space vocabulary against the floors
 * payload: unit abbreviation ("ED") → full unit name ("3 West — Medical ICU
 * (MICU)") → bed label ("MICU-01" = {ABBR}-{NN}: resolve the bed's unit and use
 * that unit's plate centroid). Anything else (free-text destinations like
 * "Main Lobby Discharge") is unresolvable → the off-map gutter.
 */
class FlowSpaceResolver(window: FlowWindowData, floors: FlowFloorsDocument?) {
    private data class UnitRef(val unitId: Int, val floor: Int)

    private val unitsByAbbr: Map<String, UnitRef>
    private val unitsByName: Map<String, UnitRef>
    private val unitPlateRect = mutableMapOf<Int, Pair<Int, List<Double>>>()
    private val bedPlateRect = mutableMapOf<Int, Pair<Int, List<Double>>>()
    private val bedUnit: Map<String, Int> // bed label → unit_id (from bed_statuses when present)

    init {
        val byAbbr = mutableMapOf<String, UnitRef>()
        val byName = mutableMapOf<String, UnitRef>()
        window.spacesFloors.forEach { floor ->
            floor.units.forEach { unit ->
                val ref = UnitRef(unit.unitId, floor.floor)
                if (unit.abbr.isNotBlank()) byAbbr.putIfAbsent(unit.abbr.lowercase(), ref)
                if (unit.name.isNotBlank()) byName.putIfAbsent(unit.name.lowercase(), ref)
            }
        }
        unitsByAbbr = byAbbr
        unitsByName = byName
        floors?.floors?.forEach { floor ->
            floor.spaces.forEach { plate ->
                when {
                    (plate.category == "unit" || plate.category == "zone") && plate.unitId != null ->
                        unitPlateRect.putIfAbsent(plate.unitId, floor.floor to plate.rect)
                    plate.category == "bed" && plate.bedId != null ->
                        bedPlateRect.putIfAbsent(plate.bedId, floor.floor to plate.rect)
                }
            }
        }
        bedUnit = window.bedStatuses
            .filter { it.unitId != null }
            .associate { it.label.lowercase() to it.unitId!! }
    }

    fun resolveSpace(raw: String?): FlowSpacePoint? {
        val s = raw?.trim()?.lowercase().orEmpty()
        if (s.isEmpty()) return null
        unitsByAbbr[s]?.let { return unitPoint(it) }
        unitsByName[s]?.let { return unitPoint(it) }
        bedUnit[s]?.let { unitId -> return resolveUnit(unitId) }
        // Bed label pattern {ABBR}-{NN}: the bed's unit owns the endpoint.
        val match = Regex("^(.+)-(\\d+)$").find(s)
        if (match != null) {
            unitsByAbbr[match.groupValues[1]]?.let { return unitPoint(it) }
        }
        return null
    }

    fun resolveUnit(unitId: Int?): FlowSpacePoint? {
        if (unitId == null) return null
        unitPlateRect[unitId]?.let { (floor, rect) -> return FlowSpacePoint(floor, rect) }
        val ref = unitsByAbbr.values.firstOrNull { it.unitId == unitId }
            ?: unitsByName.values.firstOrNull { it.unitId == unitId }
        return ref?.let { FlowSpacePoint(it.floor, null) }
    }

    fun resolveBed(bedId: Int?): FlowSpacePoint? {
        if (bedId == null) return null
        bedPlateRect[bedId]?.let { (floor, rect) -> return FlowSpacePoint(floor, rect) }
        return null
    }

    private fun unitPoint(ref: UnitRef): FlowSpacePoint {
        unitPlateRect[ref.unitId]?.let { (floor, rect) -> return FlowSpacePoint(floor, rect) }
        return FlowSpacePoint(ref.floor, null)
    }
}

/**
 * The transporter's trips at scrub time t: transport_status events at/before t
 * (the replayed record) plus transport_due projections still ahead of t (ghost
 * trips — dashed, confidence-mapped, provenance on tap).
 */
fun transportTrips(window: FlowWindowData, resolver: FlowSpaceResolver, scrubT: Long): List<TransportTrip> {
    val past = window.events
        .filter { it.kind == "transport_status" && it.tMs <= scrubT }
        .map { event ->
            TransportTrip(
                selection = FlowSelection.Event(event),
                label = event.label,
                fromRaw = event.fromSpace,
                toRaw = event.toSpace,
                from = resolver.resolveSpace(event.fromSpace),
                to = resolver.resolveSpace(event.toSpace),
                ghost = false,
                alpha = 0.85f,
                warning = event.tier == "warning" || event.tier == "critical",
            )
        }
    val ahead = window.projections
        .filter { it.kind == "transport_due" && it.tMs > scrubT }
        .map { ghost ->
            TransportTrip(
                selection = FlowSelection.Ghost(ghost),
                label = ghost.label,
                fromRaw = null,
                toRaw = null,
                from = resolver.resolveBed(ghost.bedId) ?: resolver.resolveUnit(ghost.unitId),
                to = null,
                ghost = true,
                alpha = ghost.confidenceAlpha,
                warning = false,
            )
        }
    return past + ahead
}

/** Trips with a free-text endpoint the map can't place — surfaced in the gutter. */
fun offMapTrips(trips: List<TransportTrip>): List<TransportTrip> =
    trips.filter {
        (it.from == null && !it.fromRaw.isNullOrBlank()) ||
            (it.to == null && !it.toRaw.isNullOrBlank())
    }

private fun tripColor(trip: TransportTrip): Color = when {
    trip.ghost -> Z.ink
    trip.warning -> Z.statusWarning
    else -> Z.primary
}

private fun DrawScope.drawArrowHead(tip: Offset, towards: Offset, color: Color, alpha: Float) {
    val angle = atan2(tip.y - towards.y, tip.x - towards.x)
    val len = 5.dp.toPx()
    listOf(angle + 0.5f, angle - 0.5f).forEach { a ->
        drawLine(
            color = color.copy(alpha = alpha),
            start = tip,
            end = Offset(tip.x - len * cos(a), tip.y - len * sin(a)),
            strokeWidth = 1.5.dp.toPx(),
        )
    }
}

private fun DrawScope.drawTripArc(start: Offset, end: Offset, trip: TransportTrip) {
    val color = tripColor(trip)
    val mid = Offset((start.x + end.x) / 2f, (start.y + end.y) / 2f)
    val lift = 18.dp.toPx() + (end - start).getDistance() * 0.12f
    val control = Offset(mid.x, mid.y - lift)
    val path = Path().apply {
        moveTo(start.x, start.y)
        quadraticBezierTo(control.x, control.y, end.x, end.y)
    }
    val dash = if (trip.ghost) PathEffect.dashPathEffect(floatArrayOf(5.dp.toPx(), 4.dp.toPx())) else null
    drawPath(
        path,
        color = color.copy(alpha = trip.alpha),
        style = Stroke(width = 1.5.dp.toPx(), pathEffect = dash),
    )
    drawCircle(color = color.copy(alpha = trip.alpha), radius = 2.5.dp.toPx(), center = start)
    drawArrowHead(end, control, color, trip.alpha)
}

/** Short curved stub — a trip leaving (or entering) the displayed floor. */
private fun DrawScope.drawTripStub(anchor: Offset, trip: TransportTrip, departing: Boolean) {
    val color = tripColor(trip)
    val dx = 22.dp.toPx()
    val far = Offset(anchor.x + dx, anchor.y - dx)
    val control = Offset(anchor.x + dx * 0.2f, anchor.y - dx)
    val (start, end) = if (departing) anchor to far else far to anchor
    val path = Path().apply {
        moveTo(start.x, start.y)
        quadraticBezierTo(control.x, control.y, end.x, end.y)
    }
    val dash = if (trip.ghost) PathEffect.dashPathEffect(floatArrayOf(5.dp.toPx(), 4.dp.toPx())) else null
    drawPath(
        path,
        color = color.copy(alpha = trip.alpha),
        style = Stroke(width = 1.5.dp.toPx(), pathEffect = dash),
    )
    drawCircle(color = color.copy(alpha = trip.alpha), radius = 2.5.dp.toPx(), center = start)
    drawArrowHead(end, control, color, trip.alpha)
}

/**
 * Same-floor trip arcs drawn over the FloorPlateCanvas (no pointer input — taps
 * fall through to the plates). Trips with one endpoint on this floor render as
 * stubs; ghost trips (no destination) as departure stubs at confidence alpha.
 */
@Composable
fun FloorTripArcsOverlay(
    floor: FlowFloor,
    trips: List<TransportTrip>,
    modifier: Modifier = Modifier,
) {
    val density = LocalDensity.current
    val padPx = with(density) { floorPlatePad.toPx() }
    Canvas(modifier.fillMaxSize()) {
        val transform = FlowPlanTransform(floor.bounds, size, padPx)
        fun anchor(point: FlowSpacePoint?): Offset? =
            point?.takeIf { it.floor == floor.floor }?.rect?.let(transform::center)

        trips.forEach { trip ->
            val start = anchor(trip.from)
            val end = anchor(trip.to)
            when {
                start != null && end != null -> drawTripArc(start, end, trip)
                start != null -> drawTripStub(start, trip, departing = true)
                end != null -> drawTripStub(end, trip, departing = false)
            }
        }
    }
}

/** A trip reduced to floors for the house stack; fromFloor == toFloor renders as a stub. */
data class HouseTripArc(
    val fromFloor: Int,
    val toFloor: Int,
    val color: Color,
    val dashed: Boolean,
    val alpha: Float,
)

/** House-stack arcs: cross-floor trips, plus same-/single-floor trips as stubs. */
fun houseTripArcs(trips: List<TransportTrip>): List<HouseTripArc> =
    trips.mapNotNull { trip ->
        val from = trip.from?.floor
        val to = trip.to?.floor
        val a = from ?: to ?: return@mapNotNull null
        HouseTripArc(
            fromFloor = a,
            toFloor = to ?: a,
            color = tripColor(trip),
            dashed = trip.ghost,
            alpha = trip.alpha,
        )
    }

/**
 * The "Off-map" gutter: trips whose endpoint is free text the map can't place
 * ("Main Lobby Discharge", SNFs, home health) listed as tappable chips.
 */
@Composable
fun OffMapTripGutter(
    trips: List<TransportTrip>,
    onSelect: (FlowSelection) -> Unit,
    modifier: Modifier = Modifier,
) {
    if (trips.isEmpty()) return
    Row(
        modifier
            .fillMaxWidth()
            .padding(horizontal = 12.dp)
            .horizontalScroll(rememberScrollState()),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text("Off-map", color = Z.inkMuted, fontSize = 10.sp)
        trips.forEach { trip ->
            val text = when {
                trip.from == null && !trip.fromRaw.isNullOrBlank() && trip.to == null && !trip.toRaw.isNullOrBlank() ->
                    "${trip.fromRaw} → ${trip.toRaw}"
                trip.to == null && !trip.toRaw.isNullOrBlank() -> "→ ${trip.toRaw}"
                else -> "${trip.fromRaw} →"
            }
            Box(
                Modifier
                    .heightIn(min = 48.dp)
                    .clickable { onSelect(trip.selection) },
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text,
                    color = Z.inkMuted,
                    fontSize = 11.sp,
                    maxLines = 1,
                    modifier = Modifier
                        .panel(corner = 8)
                        .padding(horizontal = 8.dp, vertical = 5.dp),
                )
            }
        }
    }
}

File: hummingbird/iosApp/Hummingbird/Features/Flow/TransportRoutesLayer.swift
import SwiftUI

// The transporter's route layer (P1 lens): completed/active trips as solid arcs between
// origin and destination, upcoming trips as dashed ghost arcs with confidence-mapped
// opacity, and an "Off-map" gutter for endpoints that name no mappable space.

// MARK: Trips

/// One trip drawn on the map: the latest transport_status at/before the scrub time
/// (solid), or a transport_due projection still ahead of it (ghost).
struct FlowTrip: Identifiable {
    let id: String
    let fromLabel: String?
    let toLabel: String?
    let ghost: Bool
    let opacity: Double
    /// STAT tier renders in the warning tint — never coral (earned urgency).
    let warning: Bool
    let unitId: Int?
    let bedId: Int?
    let time: Date?
    let selection: FlowSelection
}

enum FlowTripBuilder {
    /// Trips visible at scrub time `t`: per transport request, the latest status event at
    /// or before t (solid); plus every transport_due still due at or after t (ghost).
    static func trips(window: FlowWindowData?, at t: Date) -> [FlowTrip] {
        guard let window else { return [] }

        var latestByRef: [String: FlowTimelineEvent] = [:]
        for event in window.events where event.kind == "transport_status" {
            guard let time = event.time, time <= t else { continue }
            let key = event.entity?.ref ?? event.id
            if let current = latestByRef[key], (current.time ?? .distantPast) >= time { continue }
            latestByRef[key] = event
        }

        var trips: [FlowTrip] = latestByRef.values
            .sorted { ($0.time ?? .distantPast) < ($1.time ?? .distantPast) }
            .map { event in
                FlowTrip(id: "event|\(event.id)",
                         fromLabel: event.fromSpace, toLabel: event.toSpace,
                         ghost: false, opacity: 0.85,
                         warning: event.capacity == .warning || event.capacity == .critical,
                         unitId: event.unitId, bedId: nil, time: event.time,
                         selection: .event(event))
            }

        for projection in window.projections where projection.kind == "transport_due" {
            guard let time = projection.time, time >= t else { continue }
            let route = parseRoute(label: projection.label)
            trips.append(FlowTrip(id: "projection|\(projection.id)",
                                  fromLabel: route?.from, toLabel: route?.to,
                                  ghost: true, opacity: projection.confidenceLevel.ghostOpacity,
                                  warning: false,
                                  unitId: projection.unitId, bedId: projection.bedId, time: time,
                                  selection: .projection(projection)))
        }
        return trips
    }

    /// transport_due carries its route only in the label ("Transport due · ED → MICU-01");
    /// derived ghosts ("Likely discharge transport") have no route text at all.
    static func parseRoute(label: String?) -> (from: String, to: String)? {
        guard let tail = label?.components(separatedBy: " · ").last else { return nil }
        let parts = tail.components(separatedBy: " → ")
        guard parts.count == 2 else { return nil }
        let from = parts[0].trimmingCharacters(in: .whitespaces)
        let to = parts[1].trimmingCharacters(in: .whitespaces)
        guard !from.isEmpty, !to.isEmpty else { return nil }
        return (from, to)
    }
}

// MARK: Space resolution

/// Resolves the data plane's space vocabulary — unit abbreviation ("ED"), full unit name
/// ("3 West — Medical ICU (MICU)"), or bed label ("MICU-01") — to a floor + plate centroid.
/// Free text ("Main Lobby Discharge") stays unresolved and lands in the off-map gutter.
struct FlowSpaceResolver {
    struct Endpoint {
        let floor: Int
        /// Plate centroid in plan feet; nil when the unit has census but no geometry.
        let plan: CGPoint?
    }

    private var unitIdByAbbr: [String: Int] = [:]
    private var unitIdByName: [String: Int] = [:]
    private var floorByUnitId: [Int: Int] = [:]
    private var planByUnitId: [Int: (floor: Int, point: CGPoint)] = [:]

    init(window: FlowWindowData?, floors: FlowFloorsDocument?) {
        for floor in window?.spaces?.floors ?? [] {
            for unit in floor.units {
                floorByUnitId[unit.unitId] = floor.floor
                if let abbr = unit.abbr { unitIdByAbbr[abbr.lowercased()] = unit.unitId }
                if let name = unit.name { unitIdByName[name.lowercased()] = unit.unitId }
            }
        }
        for floor in floors?.floors ?? [] {
            for plate in floor.spaces where plate.category == "unit" {
                guard let unitId = plate.unitId, plate.rect.count >= 4 else { continue }
                planByUnitId[unitId] = (floor.floor, CGPoint(x: plate.rect[0] + plate.rect[2] / 2,
                                                            y: plate.rect[1] + plate.rect[3] / 2))
            }
        }
    }

    func endpoint(space raw: String?) -> Endpoint? {
        guard let space = raw?.trimmingCharacters(in: .whitespaces), !space.isEmpty else { return nil }
        let key = space.lowercased()
        if let unitId = unitIdByAbbr[key] ?? unitIdByName[key] {
            return endpoint(unitId: unitId)
        }
        // Bed label {ABBR}-{NN}: resolve the bed's unit, use the unit's plate centroid.
        let parts = space.split(separator: "-")
        if parts.count >= 2, Int(parts.last ?? "") != nil,
           let unitId = unitIdByAbbr[parts.dropLast().joined(separator: "-").lowercased()] {
            return endpoint(unitId: unitId)
        }
        return nil
    }

    func endpoint(unitId: Int?) -> Endpoint? {
        guard let unitId else { return nil }
        if let plan = planByUnitId[unitId] { return Endpoint(floor: plan.floor, plan: plan.point) }
        if let floor = floorByUnitId[unitId] { return Endpoint(floor: floor, plan: nil) }
        return nil
    }
}

/// A trip joined to its resolved endpoints. Origins of derived discharge ghosts (no route
/// text) fall back to the projection's bed/unit; a discharge's destination is off-map by
/// nature and simply doesn't draw.
struct ResolvedTrip: Identifiable {
    let trip: FlowTrip
    let from: FlowSpaceResolver.Endpoint?
    let to: FlowSpaceResolver.Endpoint?

    var id: String { trip.id }

    init(trip: FlowTrip, resolver: FlowSpaceResolver) {
        self.trip = trip
        from = trip.fromLabel != nil
            ? resolver.endpoint(space: trip.fromLabel)
            : resolver.endpoint(unitId: trip.unitId)
        to = resolver.endpoint(space: trip.toLabel)
    }

    /// Chip copy for the off-map gutter — only trips that NAME a space we can't map.
    var offMapText: String? {
        let fromLost = trip.fromLabel != nil && from == nil
        let toLost = trip.toLabel != nil && to == nil
        switch (fromLost, toLost) {
        case (true, true): return "\(trip.fromLabel ?? "") → \(trip.toLabel ?? "")"
        case (false, true): return "→ \(trip.toLabel ?? "")"
        case (true, false): return "\(trip.fromLabel ?? "") →"
        case (false, false): return nil
        }
    }
}

// MARK: House arcs

/// Route arcs over the house stack: solid past trips, dashed ghosts, warning tint for
/// STAT. Cross-floor trips bow between slabs; same-floor trips hump over their slab.
/// Ghosts with only an origin (derived discharge trips) draw a dashed departure ring.
struct TransportHouseArcs: View {
    let trips: [ResolvedTrip]
    let anchors: [Int: Anchor<CGRect>]
    /// Plan bounds per floor ([x, y, w, h]) for positioning endpoints along a slab.
    let boundsByFloor: [Int: [Double]]

    var body: some View {
        GeometryReader { proxy in
            let slabRects = anchors.mapValues { proxy[$0] }
            Canvas { context, _ in
                for (index, resolved) in trips.enumerated() {
                    draw(resolved, index: index, slabs: slabRects, in: &context)
                }
            }
        }
        .allowsHitTesting(false) // arcs annotate; slabs stay tappable underneath
    }

    private func draw(_ resolved: ResolvedTrip, index: Int,
                      slabs: [Int: CGRect], in context: inout GraphicsContext) {
        let trip = resolved.trip
        let color = trip.warning ? Z.status(.warning) : (trip.ghost ? Z.ink : Z.primary)
        let shading = GraphicsContext.Shading.color(color.opacity(trip.opacity))
        let style = trip.ghost
            ? StrokeStyle(lineWidth: 1.5, dash: [4, 3])
            : StrokeStyle(lineWidth: 1.5)

        guard let from = resolved.from, let fromSlab = slabs[from.floor] else { return }
        let start = point(for: from, in: fromSlab, defaultFraction: 0.3)

        guard let to = resolved.to, let toSlab = slabs[to.floor] else {
            // Origin-only ghost (derived discharge trip): a dashed departure ring.
            if trip.ghost {
                let ring = CGRect(x: start.x - 4, y: start.y - 4, width: 8, height: 8)
                context.stroke(Path(ellipseIn: ring), with: shading, style: style)
            }
            return
        }
        let end = point(for: to, in: toSlab, defaultFraction: 0.7)

        var path = Path()
        path.move(to: start)
        if from.floor == to.floor {
            // Same-floor: hump above the slab.
            path.addQuadCurve(to: end, control: CGPoint(x: (start.x + end.x) / 2, y: start.y - 18))
        } else {
            // Cross-floor: bow sideways; stagger overlapping trips by index.
            let bow: CGFloat = 28 + CGFloat(index % 3) * 14
            path.addQuadCurve(to: end,
                              control: CGPoint(x: max(start.x, end.x) + bow, y: (start.y + end.y) / 2))
        }
        context.stroke(path, with: shading, style: style)

        // Destination marker: filled for real trips, open for ghosts.
        let dot = CGRect(x: end.x - 3, y: end.y - 3, width: 6, height: 6)
        if trip.ghost {
            context.stroke(Path(ellipseIn: dot), with: shading, lineWidth: 1.5)
        } else {
            context.fill(Path(ellipseIn: dot), with: shading)
        }
    }

    /// Endpoint x along the slab from its plan-x fraction when geometry exists;
    /// otherwise a fixed origin/destination station.
    private func point(for endpoint: FlowSpaceResolver.Endpoint,
                       in slab: CGRect, defaultFraction: CGFloat) -> CGPoint {
        var fraction = defaultFraction
        if let plan = endpoint.plan, let bounds = boundsByFloor[endpoint.floor],
           bounds.count >= 4, bounds[2] > 0 {
            fraction = min(max((plan.x - bounds[0]) / bounds[2], 0.1), 0.9)
        }
        return CGPoint(x: slab.minX + fraction * slab.width, y: slab.midY)
    }
}

// MARK: Off-map gutter

/// Trips whose named origin/destination isn't a mappable space (lobbies, SNFs, home
/// health) — listed as chips under the map instead of silently dropped.
struct OffMapGutter: View {
    let trips: [ResolvedTrip]
    let onSelect: (FlowSelection) -> Void

    private var items: [(id: String, text: String, trip: ResolvedTrip)] {
        trips.compactMap { resolved in
            resolved.offMapText.map { (resolved.id, $0, resolved) }
        }
    }

    var body: some View {
        if !items.isEmpty {
            VStack(alignment: .leading, spacing: Z.s2) {
                Text("OFF-MAP")
                    .font(.system(size: 11, weight: .semibold)).tracking(0.5)
                    .foregroundStyle(Z.inkMuted)
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: Z.s2) {
                        ForEach(items, id: \.id) { item in
                            chip(item.text, trip: item.trip)
                        }
                    }
                }
            }
        }
    }

    private func chip(_ text: String, trip: ResolvedTrip) -> some View {
        Button {
            onSelect(trip.trip.selection)
        } label: {
            HStack(spacing: Z.s1) {
                Image(systemName: trip.trip.ghost ? "clock" : "figure.walk")
                    .font(.system(size: 10, weight: .medium))
                Text(text)
                    .font(.system(size: 12, weight: .medium))
                    .lineLimit(1)
            }
            .foregroundStyle(trip.trip.warning ? Z.status(.warning) : Z.inkMuted)
            .padding(.horizontal, Z.s3)
            .frame(height: 44) // full-height touch target even though the capsule reads small
            .contentShape(Capsule())
        }
        .buttonStyle(.plain)
        .background(
            Capsule().fill(Z.surface)
                .padding(.vertical, Z.s2)
        )
        .overlay(
            Capsule()
                .strokeBorder(Z.border, style: trip.trip.ghost
                    ? StrokeStyle(lineWidth: 1, dash: [4, 3])
                    : StrokeStyle(lineWidth: 1))
                .padding(.vertical, Z.s2)
        )
        .accessibilityLabel("Off-map trip, \(text)")
    }
}
